File: samples/Demo/Fila/Widgets/RevenueChartWidget.cs
using Demo.Data;
using Fila.Widgets;
using Microsoft.EntityFrameworkCore;

namespace Demo.Fila.Widgets;

/// <summary>Revenue per day over the last two weeks — the phase's one chart, driven by a real
/// query rather than sample points.</summary>
public sealed class RevenueChartWidget(AppDb db) : ChartWidget
{
    private const int Days = 14;

    public override string? Heading => "Revenue";

    public override string? Description => $"Last {Days} days";

    protected override async Task<ChartData> GetDataAsync(WidgetContext context)
    {
        // Inclusive of today, so the series always ends on "now" rather than on the last day
        // that happened to have an order.
        var firstDay = DateTime.UtcNow.Date.AddDays(-(Days - 1));

        // The date filter runs in the database; the grouping and the sum do not. Order.Total is
        // decimal, which EF's SQLite provider stores as TEXT — GROUP BY with SUM() over that is
        // a string aggregate, not an arithmetic one. Bounded by the 14-day window either way.
        var recent = await db.Orders
            .Where(o => o.CreatedAt >= firstDay)
            .Select(o => new { o.CreatedAt, o.Total })
            .ToListAsync(context.Ct);

        var byDay = recent
            .GroupBy(o => o.CreatedAt.Date)
            .ToDictionary(g => g.Key, g => g.Sum(o => o.Total));

        // Every day in the window gets a point, including the ones with no orders. Plotting only
        // the days that have data would silently compress the gaps and misstate the trend.
        var points = Enumerable.Range(0, Days)
            .Select(offset =>
            {
                var day = firstDay.AddDays(offset);
                return new ChartPoint(day.ToString("MMM d"), byDay.GetValueOrDefault(day));
            })
            .ToList();

        return new ChartData(points, ValuePrefix: "$");
    }
}

File: samples/FilamentReference/app/Filament/Widgets/RevenueChartWidget.php
<?php

namespace App\Filament\Widgets;

use App\Models\Order;
use Carbon\Carbon;
use Filament\Support\RawJs;
use Filament\Widgets\ChartWidget;

class RevenueChartWidget extends ChartWidget
{
    private const DAYS = 14;

    protected ?string $heading = 'Revenue';

    protected ?string $description = 'Last '.self::DAYS.' days';

    protected function getData(): array
    {
        $firstDay = Carbon::today()->subDays(self::DAYS - 1);

        $byDay = Order::where('created_at', '>=', $firstDay)
            ->get(['created_at', 'total'])
            ->groupBy(fn (Order $order) => $order->created_at->toDateString())
            ->map(fn ($group) => $group->sum('total'));

        $labels = [];
        $data = [];

        foreach (range(0, self::DAYS - 1) as $offset) {
            $day = $firstDay->copy()->addDays($offset);
            $labels[] = $day->format('M j');
            $data[] = (float) ($byDay[$day->toDateString()] ?? 0);
        }

        return [
            'datasets' => [
                [
                    'label' => 'Revenue',
                    'data' => $data,
                ],
            ],
            'labels' => $labels,
        ];
    }

    protected function getType(): string
    {
        return 'line';
    }

    protected function getOptions(): array|RawJs|null
    {
        return RawJs::make(<<<'JS'
            {
                scales: {
                    y: {
                        ticks: {
                            callback: (value) => '$' + value,
                        },
                    },
                },
            }
            JS);
    }
}
